### PR TITLE
fix: massage release notes links

### LINKED ISSUE #17404: matchAll in collectLinkPosition function ignores colors number when getting startOffset
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest 

### Please select which platform you are using if self-hosting.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

The problem is when running
https://github.com/renovatebot/renovate/blob/8f8f059d4fdc899a3f4cee805356389d67672561/lib/modules/platform/github/massage-markdown-links.ts#L39

the tree.value will be : pnpm rebuild accepts --store-dir by @​chengcyber in https://github.com/pnpm/pnpm/pull/5036

image

but if I copy this value and check it in regex101 I see that there are some characters does not appear in `tree.value`(in this case "U+200B")

image

so when getting the `startOffset`, there are 6 characters missing and the number will be `startOffset-6`.

please see `what's changed` and `New Contributors` sections:
reproduction repo: https://github.com/MaronHatoum/16921/pull/24

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by MaronHatoum**
## Changes
added new regex to massage `What's Changed` and `New Contributors` in release notes.
before the fix, these sections did not add a link to the user's profile, and the PRs links have been displayed incorrectly.
- please take a look at Release Notes `What's Changed` and `New Contributors` sections in the PRs below:
**Before the fix:**
https://github.com/MaronHatoum/16921/pull/43
**After the fix:**
https://github.com/MaronHatoum/16921/pull/48

## Context
closes https://github.com/renovatebot/renovate/issues/17404

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by MaronHatoum** [comment]
- Release notes before the fix (check **What's Changed** section):
https://github.com/MaronHatoum/16921/pull/43

- Release notes after the fix (check **What's Changed** section):
https://github.com/MaronHatoum/16921/pull/45

File: lib/modules/platform/github/massage-markdown-links.spec.ts [hunk 1]
```diff
@@ -87,4 +87,16 @@ describe('modules/platform/github/massage-markdown-links', () => {
...
   );
+
+  it('replace url contains disappear text inline (colors)', () => {
+    const input = [
+      'pnpm rebuild accepts --store-dir by @&#8203;user in https://github.com/foo/bar/issues/1',
+    ].join('\n');
+    const res = massageMarkdownLinks(input);
+    expect(res).toEqual(
+      [
+        'pnpm rebuild accepts --store-dir by @user in [https://github.com/foo/bar/issues/1](https://togithub.com/foo/bar/issues/1)',
```
Comment (HonkingGoose) [inline lib/modules/platform/github/massage-markdown-links.spec.ts:98]: Your **after fix** PR is pinging _all_ the authors mentioned in the release notes. 🙈

I suspect you can fix this by putting the `@user` parts between backticks or a `<code></code>` block, like this:

```ts
        'pnpm rebuild accepts --store-dir by `@user` in [https://github.com/foo/bar/issues/1](https://togithub.com/foo/bar/issues/1)',
```

**Bad (current behavior):**

> blabla by @HonkingGoose in [link to PR](#)

**Correct:**

> blabla by `@HonkingGoose` in [link to PR](#)

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
I like this idea, but there's a problem. 😉

File: lib/modules/platform/github/massage-markdown-links.spec.ts (hunk 1 above)
Comment (MaronHatoum) [reply lib/modules/platform/github/massage-markdown-links.spec.ts]: I added backticks around @user but there is no link to the user profile, is it the correct behavior?
**result:**
https://github.com/MaronHatoum/16921/pull/45

Comment (viceice) [reply lib/modules/platform/github/massage-markdown-links.spec.ts]: Looks like there is still one wrong username
image

File: lib/modules/platform/github/massage-markdown-links.ts [hunk 2]
```diff
@@ -58,12 +58,19 @@ export function massageMarkdownLinks(content: string): string {
   try {
     const rightSpaces = content.replace(content.trimEnd(), '');
     const matches: UrlMatch[] = [];
-    remark().use(collectLinkPosition(content, matches)).processSync(content);
+    //const newContent = content.replace(regEx(/@&#8203;/g), '@');
+    const newContent = content.replace(
+      regEx(/@&#8203;(?<name>[a-z1-9]*)/g),
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.ts:63]: needs to be case insensitive
```suggestion
      regEx(/@&#8203;(?<name>[a-z1-9]*)/gi),
```
or


```suggestion
      regEx(/@&#8203;(?<name>[A-Za-z1-9]*)/g),
```

File: lib/modules/platform/github/massage-markdown-links.spec.ts (hunk 1 above)
Comment (MaronHatoum) [reply lib/modules/platform/github/massage-markdown-links.spec.ts]: can you check now please:
https://github.com/MaronHatoum/16921/pull/46

File: lib/modules/platform/github/massage-markdown-links.spec.ts [hunk 3]
```diff
@@ -87,4 +87,18 @@ describe('modules/platform/github/massage-markdown-links', () => {
...
+
+  it('replace url contains disappear text inline (colors)', () => {
+    const input = [
+      'pnpm rebuild accepts --store-dir by @&#8203;user in https://github.com/foo/bar/issues/1\n' +
+        'pnpm rebuild accepts --store-dir by @&#8203;UsEr in https://github.com/foo/bar/issues/2',
+    ].join('\n');
+    const res = massageMarkdownLinks(input);
+    expect(res).toEqual(
+      [
+        'pnpm rebuild accepts --store-dir by `@user` in [https://github.com/foo/bar/issues/1](https://togithub.com/foo/bar/issues/1)\n' +
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.spec.ts:99]:
```suggestion
        'pnpm rebuild accepts --store-dir by `@user` in [foo/bar#1](https://togithub.com/foo/bar/issues/1)\n' +
```

we should do this in a follow up PR to reduce size some more

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/platform/github/massage-markdown-links.spec.ts (hunk 3 above)
Comment (MaronHatoum) [reply lib/modules/platform/github/massage-markdown-links.spec.ts:99]: @viceice  https://github.com/MaronHatoum/16921/pull/46

File: lib/modules/platform/github/massage-markdown-links.ts [hunk 4]
```diff
@@ -58,12 +68,18 @@ export function massageMarkdownLinks(content: string): string {
   try {
     const rightSpaces = content.replace(content.trimEnd(), '');
     const matches: UrlMatch[] = [];
-    remark().use(collectLinkPosition(content, matches)).processSync(content);
+    const newContent = content.replace(
+      regEx(/@&#8203;(?<name>[A-Za-z1-9]*)/gi),
```
Comment (PhilipAbed) [inline lib/modules/platform/github/massage-markdown-links.ts:72]: regex should match everything that is not space 
example name: `@philip-abed` will not be caught here.

**Comment by PhilipAbed** [comment]
This special character is added by us at 
https://github.com/renovatebot/renovate/blob/ae755d488078953f6fed43eec5a008590f9ade9f/lib/util/markdown.ts#L6-L26

With the changes in this PR, we have a code that adds something in one place, and then removes it in another place.

Instead we might be able to do it the correct way from the start, WDYT?

File: lib/modules/platform/github/massage-markdown-links.spec.ts [hunk 5]
```diff
@@ -87,4 +87,20 @@ describe('modules/platform/github/massage-markdown-links', () => {
...
+  it('replace url contains disappear text inline', () => {
+    const input = [
+      'pnpm rebuild accepts --store-dir by @&#8203;user in https://github.com/foo/bar/issues/1\n' +
+        'pnpm rebuild accepts --store-dir by @&#8203;UsEr in https://github.com/foo/bar/issues/2\n' +
+        'pnpm rebuild accepts --store-dir by @&#8203;user-name in https://github.com/foo/bar/issues/3',
+    ].join('\n');
+    const res = massageMarkdownLinks(input);
+    expect(res).toEqual(
+      [
+        'pnpm rebuild accepts --store-dir by `@user` in [foo/bar#1](https://togithub.com/foo/bar/issues/1)\n' +
```
Comment (HonkingGoose) [inline lib/modules/platform/github/massage-markdown-links.spec.ts:100]: The upstream changelog links to the user profiles, so we should try to keep the links in. 😉 

### Link to user profile via `github.com`

Maybe this still causes pings?

```markdown
[`@HonkingGoose`](https://github.com/HonkingGoose)
```

### Link to user profile via `togithub.com` redirect

Or we could replace the `github.com` domain with our `togithub.com` domain to be sure we don't ping back:

```markdown
[`@HonkingGoose`](https://togithub.com/HonkingGoose)
```

Comment (MaronHatoum) [reply lib/modules/platform/github/massage-markdown-links.spec.ts:100]: @HonkingGoose: https://github.com/MaronHatoum/16921/pull/48

File: lib/modules/platform/github/massage-markdown-links.ts [hunk 6]
```diff
@@ -58,12 +68,18 @@ export function massageMarkdownLinks(content: string): string {
   try {
     const rightSpaces = content.replace(content.trimEnd(), '');
     const matches: UrlMatch[] = [];
-    remark().use(collectLinkPosition(content, matches)).processSync(content);
+    const newContent = content.replace(
+      regEx(/@&#8203;(?<name>[a-zA-Z1-9-!$%^&*()_+|~=`{}[\]:";'<>?,.]*)/gi),
+      '[`@$<name>`](https://togithub.com/$<name>)'
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.ts:73]: @HonkingGoose do we still need the code trick when we manually link the user?

Comment (HonkingGoose) [reply lib/modules/platform/github/massage-markdown-links.ts:73]: I don't know. Let's try removing the `code` trick, and see what happens then.

```suggestion
      '[@$<name>](https://togithub.com/$<name>)'
```

**Comment by rarkins** [comment]
Why do we need to arrange release notes? I don't understand what the issue is about

**Comment by MaronHatoum** [comment]
> Why do we need to arrange release notes? I don't understand what the issue is about

sorry, I will add a clearer description.

**Comment by MaronHatoum** [comment]
> Why do we need to arrange release notes? I don't understand what the issue is about

added more description.

**Comment by rarkins** [comment]
It looks like your PRs have been spamming pnpm, Eg https://github.com/pnpm/pnpm/issues/5071

Has this been fixed?

**Comment by MaronHatoum** [comment]
> It looks like your PRs have been spamming pnpm, Eg [pnpm/pnpm#5071](https://github.com/pnpm/pnpm/issues/5071)
> 
> Has this been fixed?

yes

**Comment by rarkins** [comment]
I've labeled this PR as large impact, because if there's any mistake or misunderstanding here with this parsing then it could result in mass backlink spam. I'm not sure of the best way to monitor for this problem in production though because I'm not aware of any github.com search we can do to find our own backlinks. i.e. in a worst case scenario we could be backlink spamming for hours or days before we annoy someone enough to report it.

File: lib/modules/platform/github/massage-markdown-links.ts [hunk 7]
```diff
@@ -13,11 +13,17 @@ interface UrlMatch {
 
 const urlRegex =
   /(?:https?:)?(?:\/\/)?(?:www\.)?(?<!api\.)(?:to)?github\.com\/[-_a-z0-9]+\/[-_a-z0-9]+\/(?:discussions|issues|pull)\/[0-9]+(?:#[-_a-z0-9]+)?/i; // TODO #12872 (?<!re) after text not matching
+const reduceUrlRegex =
+  /(?:https?:)?(?:\/\/)?(?:www\.)?github\.com\/(?<reg>[a-zA-Z1-9-!$%^&*()_+|~=`{}[\]:";'<>?,.]*)\/(?<repo>[a-zA-Z1-9-!$%^&*()_+|~=`{}[\]:";'<>?,.]*)\/([a-zA-Z1-9-!$%^&*()_+|~=`{}[\]:";'<>?,.]*)\/(?<number>[\d]+)/g;
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.ts:17]: are you sure all those chars are allowed for org or repo name? the regex above is clear more strict, so both should match.

also what is `reg`? shouldn't it `org` for organization?

File: lib/modules/platform/github/massage-markdown-links.ts [hunk 8]
```diff
@@ -58,12 +68,18 @@ export function massageMarkdownLinks(content: string): string {
   try {
     const rightSpaces = content.replace(content.trimEnd(), '');
     const matches: UrlMatch[] = [];
-    remark().use(collectLinkPosition(content, matches)).processSync(content);
+    const newContent = content.replace(
+      regEx(/@&#8203;(?<name>[a-zA-Z1-9-!$%^&*()_+|~=`{}[\]:";'<>?,.]*)/gi),
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.ts:72]: same here, i don't think all those special chars are needed, as they aren't allowed by GitHub

Comment (MaronHatoum) [reply lib/modules/platform/github/massage-markdown-links.ts:72]: you are right 
the allowed chars are `_-.`

**Comment by MaronHatoum** [comment]
I have checked the fix on a real repository.
if you take a look to this PR(after the fix) which has been created 11 days ago https://github.com/MaronHatoum/16921/pull/49
and then you check the URLs in release notes you can see that the last "ping/spam" was 13 days ago which means that the fix I did here stopped spamming related `PRs/discussions/issues` that were mentioned in PR release notes.

File: lib/modules/platform/github/massage-markdown-links.ts [hunk 9]
```diff
@@ -13,11 +13,17 @@ interface UrlMatch {
 
 const urlRegex =
   /(?:https?:)?(?:\/\/)?(?:www\.)?(?<!api\.)(?:to)?github\.com\/[-_a-z0-9]+\/[-_a-z0-9]+\/(?:discussions|issues|pull)\/[0-9]+(?:#[-_a-z0-9]+)?/i; // TODO #12872 (?<!re) after text not matching
+const reduceUrlRegex =
+  /(?:https?:)?(?:\/\/)?(?:www\.)?github\.com\/(?<org>[a-zA-Z1-9\-_.]*)\/(?<repo>[a-zA-Z1-9\-_.]*)\/([a-zA-Z1-9\-_.]*)\/(?<number>[\d]+)/g;
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.ts:17]: please align both to same allowed chars

Comment (MaronHatoum) [reply lib/modules/platform/github/massage-markdown-links.ts:17]: but if you mean `urlRegex` and `reduceUrlRegex` we have a PR for this change:
https://github.com/renovatebot/renovate/pull/17819/files

File: lib/modules/platform/github/massage-markdown-links.ts [hunk 10]
```diff
@@ -57,12 +67,18 @@ export function massageMarkdownLinks(content: string): string {
   try {
     const rightSpaces = content.replace(content.trimEnd(), '');
     const matches: UrlMatch[] = [];
-    remark().use(collectLinkPosition(content, matches)).processSync(content);
+    const newContent = content.replace(
+      regEx(/@&#8203;(?<name>[a-zA-Z1-9\-_.]*)/gi),
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.ts:71]:
```suggestion
      regEx(/@&#8203;(?<name>[a-z1-9\-_.]*)/gi),
```
simplify if case insensitive flag is used.

File: lib/modules/platform/github/massage-markdown-links.spec.ts [hunk 11]
```diff
@@ -59,19 +59,19 @@ describe('modules/platform/github/massage-markdown-links', () => {
...
-    ${'https://github.com/foo/bar/issues/1'}                                                  | ${'[https://github.com/foo/bar/issues/1](https://togithub.com/foo/bar/issues/1)'}
-    ${'https://github.com/foo/bar/pull/1'}                                                    | ${'[https://github.com/foo/bar/pull/1](https://togithub.com/foo/bar/pull/1)'}
-    ${'https://github.com/foo/bar/discussions/1#comment-123'}                                 | ${'[https://github.com/foo/bar/discussions/1#comment-123](https://togithub.com/foo/bar/discussions/1#comment-123)'}
-    ${'https://github.com/foo/bar/issues/1#comment-123'}                                      | ${'[https://github.com/foo/bar/issues/1#comment-123](https://togithub.com/foo/bar/issues/1#comment-123)'}
-    ${'https://github.com/foo/bar/pull/1#comment-123'}                                        | ${'[https://github.com/foo/bar/pull/1#comment-123](https://togithub.com/foo/bar/pull/1#comment-123)'}
+    ${'github.com/foo/bar/discussions/1'}                                                     | ${'[foo/bar#1](togithub.com/foo/bar/discussions/1)'}
```
Comment (viceice) [inline lib/modules/platform/github/massage-markdown-links.spec.ts:62]: do those short links also work for discussions?

Comment (PhilipAbed) [reply lib/modules/platform/github/massage-markdown-links.spec.ts:62]: the answer is yes, but i will close this PR too and refactor it

Comment (PhilipAbed) [reply lib/modules/platform/github/massage-markdown-links.spec.ts:62]: https://github.com/renovatebot/renovate/pull/17819 is doing this refactor, this PR is an aftermath for just one case

**Pull request closed without merging**